#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import median_absolute_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import max_error
import urllib.request
from PIL import Image

## Data

In [4]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/train 2.csv", index_col = 0)
test = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 2.csv", index_col = 0)


## Data exploration

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 912 entries, 0 to 911
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Inches            912 non-null    float64
 1   Ram               912 non-null    int64  
 2   Weight            912 non-null    float64
 3   Memory            912 non-null    int64  
 4   ScreenResolution  912 non-null    int64  
 5   Price_euros       912 non-null    float64
 6   Cpu_prosesor_num  912 non-null    int64  
 7   Company_num       912 non-null    int64  
 8   TypeName_num      912 non-null    int64  
dtypes: float64(3), int64(6)
memory usage: 71.2 KB


In [6]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 391 entries, 0 to 390
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Inches            391 non-null    float64
 1   Ram               391 non-null    int64  
 2   Weight            391 non-null    float64
 3   Memory            391 non-null    int64  
 4   ScreenResolution  391 non-null    int64  
 5   Cpu_prosesor_num  391 non-null    int64  
 6   Company_num       391 non-null    int64  
 7   TypeName_num      391 non-null    int64  
dtypes: float64(2), int64(6)
memory usage: 27.5 KB


In [7]:
train.head()

Inches  Ram  Weight  Memory  ScreenResolution  Price_euros  \
0    15.6    4    2.40     500           1049088        387.0   
1    13.3    8    1.23     256           5760000       1379.0   
2    17.3    8    3.00     256           2073600        854.0   
3    12.5    8    1.26     512           2073600       1483.0   
4    17.3    8    2.80    1024           1440000        519.0   

   Cpu_prosesor_num  Company_num  TypeName_num  
0                 3            0             3  
1                15            4             4  
2                14            0             3  
3                15            7             2  
4                 2           10             3

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [8]:
X = train.drop("Price_euros", axis = 1)
y = train['Price_euros']
X.shape

(912, 8)

In [9]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [10]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

### 3. Asignar el modelo (vacío) a una variable


In [11]:
lm = LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [12]:
lm.fit(X, y)

LinearRegression()

In [13]:
coeff_df = pd.DataFrame(lm.coef_,
                        X.columns,
                        columns=['Coefficient'])
coeff_df.sort_values('Coefficient', ascending=False)

Coefficient
Weight             151.316447
Ram                 82.485667
Cpu_prosesor_num    44.930025
TypeName_num        35.594446
Company_num         15.617218
ScreenResolution     0.000110
Memory              -0.132621
Inches             -75.111574

In [14]:
predictions = lm.predict(X)
predictions

array([ 129.39477484, 1642.79331626, 1061.71414264, 1243.66427655,
        476.91991337, 1872.88041061, 1224.1131091 , 1039.06217374,
       1597.86329101, 1078.9083837 , 1642.80286788,  704.04691971,
       1597.86329101, 1154.31131914,  792.72745291, 1421.62684687,
       1056.88689994,  264.1848506 , 1018.6147664 , 1114.17472299,
       2464.72970939, 1836.87520682, 1025.87032037, 1185.75811792,
       1266.01734953,  556.53899181, 1323.06871452, 1159.48668055,
       1211.59932915,  770.10536148, 1201.47263083, 1166.26083971,
        669.52622441, 1809.24869377, 1933.99104696, 1902.28428822,
       1299.30642375,  795.69792362,  548.43173796, 1590.77141613,
        210.37363988, 1969.01667649, 3122.69496429,  337.48202846,
        455.34152698, 1363.15617304,  480.92498053, 1076.65804561,
        581.74143429,  449.79315812, 1467.97778977,  612.99480119,
       1754.90625862,  669.52622441, 1193.05870063, 2162.74360267,
        941.57192594, 1306.87279022, 1321.15632707, 1261.47731

In [15]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error ,mean_squared_error, r2_score

print("MAE: ", mean_absolute_error(y, predictions))
print("MAPE: ", mean_absolute_percentage_error(y,predictions)) # Es el complemento del R2
print("MSE: ", mean_squared_error(y,predictions))
print("RMSE: ", np.sqrt(mean_squared_error(y,predictions))) #es el mas/menos en la misma unidad que el target
print("R2: ", r2_score(y,predictions))

MAE:  282.00871371303896
MAPE:  0.2931324052296071
MSE:  153123.115190815
RMSE:  391.3094877342166
R2:  0.6700611770962484


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [16]:
# from sklearn.preprocessing import PolynomialFeatures
# poly_feats = PolynomialFeatures(degree = 3)
# poly_feats.fit(X)
# X_poly = poly_feats.transform(X)
# X_poly




# pol_reg = LinearRegression()
# pol_reg.fit(X_train, y_train)


# predictions = pol_reg.predict(X_test)


# df_preds = pd.DataFrame(predictions, columns = ['predictions'])
# df_preds['real_value'] = pd.Series(y_test)

# df_preds

# from sklearn.metrics import r2_score
# r2_score(pol_reg.predict(X_test), y_test)

### 2. Carga los datos de `test.csv` para predecir.

In [17]:
# X_pred = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 2.csv", index_col = 0)
X_pred = test
X_pred.head()

Inches  Ram  Weight  Memory  ScreenResolution  Cpu_prosesor_num  \
0    14.0    8    1.25     256           2073600                15   
1    14.0    4    1.40      32           2304000                 6   
2    14.0    2    1.40      32           2073600                 6   
3    15.6    8    2.65    1024           2073600                15   
4    11.6    2    1.10      32           1049088                 7   

   Company_num  TypeName_num  
0           14             3  
1           11             3  
2           11             3  
3            4             1  
4            2             0

In [18]:
X_pred.shape

(391, 8)

**¿Por qué da error?** 

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [19]:
X_pred.head()

Inches  Ram  Weight  Memory  ScreenResolution  Cpu_prosesor_num  \
0    14.0    8    1.25     256           2073600                15   
1    14.0    4    1.40      32           2304000                 6   
2    14.0    2    1.40      32           2073600                 6   
3    15.6    8    2.65    1024           2073600                15   
4    11.6    2    1.10      32           1049088                 7   

   Company_num  TypeName_num  
0           14             3  
1           11             3  
2           11             3  
3            4             1  
4            2             0

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [20]:
predictions_submit = lm.predict(X_pred)
predictions_submit

array([1308.34963033,  604.93144663,  414.61826064, 1070.80034562,
        234.39605617,  800.87686604,  509.8367445 , 1798.44743468,
       2033.049477  , 1154.14354018, 1155.53640136, 1285.84216169,
        969.62969974, 1141.14887453,  805.20498584, 1152.51007242,
       1234.29734032, 2016.66317279,  681.48529847, 1196.91295156,
       1308.43511714, 1017.07235606,  669.44911533, 1050.41545099,
        370.44221399, 1442.12475955,  870.89002739, 1027.39241411,
         21.79390212, 1154.14354018, 1074.61444776, 1023.37822939,
       1119.1276036 , 1007.740347  , 1325.72956233, 1330.6644425 ,
       1071.58811882, 1398.33003885,  805.86993516, 2084.03143756,
       1832.4128225 , 1109.21351493,  542.86426589, 1434.1014715 ,
       1191.85234865, 1879.24987357,  829.58251927, 1635.70144138,
       1144.36039031, 1698.39136947,  861.32667819,  706.58996744,
       1079.94949434, 1158.56273031,  783.25163665,  852.83069571,
       1094.08187023,  390.83042034, 1249.60846838, 1398.75684

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [21]:
sample = pd.read_csv("data/sample_submission.csv")

In [22]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [23]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [24]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [25]:
submission.head()

laptop_ID  Price_euros
0        750  1308.349630
1        726   604.931447
2        633   414.618261
3        363  1070.800346
4        319   234.396056

In [26]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [27]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [28]:
chequeator(submission)

You're ready to submit!
